### ISSUE TITLE
[BUG] Windows Build Error: pkg-config package missing

### ISSUE THREAD

**Issue by sciguyryan** [gtk-rs/gtk4-rs#734]
**Bug description**

I have followed the guide for setting up GTK on Windows (here: https://github.com/gtk-rs/gtk4-rs/blob/master/book/src/installation_windows.md)

Having done this, the build still fails with a message like this: 

"Most likely, you need to install a pkg-config package for your OS."

Except I have it installed, all of the path variables are set correctly and so on.

[**Backtrace**]

Any help fixing this would be greatly appreciated.

**Comment by sdroege** [comment]
If you run `pkg-config --modversion glib-2.0` in a terminal, does that work?

**Comment by sdroege** [comment]
Can you give the output of `cargo check -vvvvv` from that same console?

**Comment by sciguyryan** [comment]
> Can you give the output of `cargo check -vvvvv` from that same console?

Right, this is fascinating.

When I run the command directly in the VS Code console, I get an error:

`pkg-config : The term 'pkg-config' is not recognized as the name of a cmdlet, function, script file, or operable program. Check the spelling of the name, or if a path was included, verify that the path is correct and try again.
At line:1 char:1
+ pkg-config.exe --modversion glib-2.0
+ ~~~~~~~~~~
    + CategoryInfo          : ObjectNotFound: (pkg-config:String) [], CommandNotFoundException
    + FullyQualifiedErrorId : CommandNotFoundException`

Yet when I run it from the command line directly, no such problem. I'm at a loss as to why this problem, it seems that VS Code isn't obeying the environmental variables, which is causing things not to correctly build?

**Comment by sciguyryan** [comment]
The following is (part) of the output from the cargo check command:

       Fresh field-offset v0.3.4
       Fresh image v0.23.14
    Checking psistega3-core v0.1.0 (D:\GitHub\PsiStega3\psistega3-core)
     Running `set CARGO=\\?\C:\Users\Ryan\.rustup\toolchains\stable-x86_64-pc-windows-msvc\bin\cargo.exe&& set CARGO_CRATE_NAME=psistega3_core&& set CARGO_MANIFEST_DIR=D:\GitHub\PsiStega3\psistega3-core&& set CARGO_PKG_AUTHORS="Ryan Jones-Ward <sciguyryan@gmail.com>"&& set CARGO_PKG_DESCRIPTION="A core crate for PsiStega3."&& set CARGO_PKG_HOMEPAGE=https://github.com/sciguyryan/PsiStega3&& set CARGO_PKG_LICENSE="LGPL 2.1"&& set CARGO_PKG_LICENSE_FILE=""&& set CARGO_PKG_NAME=psistega3-core&& set CARGO_PKG_REPOSITORY=https://github.com/sciguyryan/PsiStega3&& set CARGO_PKG_VERSION=0.1.0&& set CARGO_PKG_VERSION_MAJOR=0&& set CARGO_PKG_VERSION_MINOR=1&& set CARGO_PKG_VERSION_PATCH=0&& set CARGO_PKG_VERSION_PRE=""&& set CARGO_PRIMARY_PACKAGE=1&& set PATH="D:\GitHub\PsiStega3\target\debug\deps;C:\Users\Ryan\.rustup\toolchains\stable-x86_64-pc-windows-msvc\bin;C:\Users\Ryan\.cargo\bin;C:\Users\Ryan\.rustup\toolchains\stable-x86_64-pc-windows-msvc\bin;C:\Program Files\Microsoft\jdk-17.0.1.12-hotspot\bin;C:\Program Files\Python310\Scripts\;C:\Program Files\Python310\;C:\Program Files\ImageMagick-7.1.0-Q16-HDRI;C:\Program Files\Python\Python39\Scripts\;C:\Program Files\Python\Python39\;C:\Windows\system32;C:\Windows;C:\Windows\System32\Wbem;C:\Windows\System32\WindowsPowerShell\v1.0\;C:\Windows\System32\OpenSSH\;C:\Program Files\Microsoft VS Code\bin;C:\Program Files\dotnet\;C:\Program Files\Git\cmd;C:\Program Files (x86)\Windows Kits\10\Windows Performance Toolkit\;C:\Users\Ryan\.cargo\bin;C:\Users\Ryan\AppData\Local\Microsoft\WindowsApps;C:\Users\Ryan\AppData\Local\GitHubDesktop\bin;C:\Users\Ryan\AppData\Local\Programs\MiKTeX\miktex\bin\x64\\"&& rustc --crate-name psistega3_core --edition=2021 psistega3-core\src\lib.rs --error-format=json --json=diagnostic-rendered-ansi --crate-type lib --emit=dep-info,metadata -C embed-bitcode=no -C debuginfo=2 -C metadata=bc68478ec5410774 -C extra-filename=-bc68478ec5410774 --out-dir D:\GitHub\PsiStega3\target\debug\deps -C incremental=D:\GitHub\PsiStega3\target\debug\incremental -L dependency=D:\GitHub\PsiStega3\target\debug\deps --extern aes_gcm=D:\GitHub\PsiStega3\target\debug\deps\libaes_gcm-5264815bf377079a.rmeta --extern argon2=D:\GitHub\PsiStega3\target\debug\deps\libargon2-60b2bbdff02c228e.rmeta --extern base64=D:\GitHub\PsiStega3\target\debug\deps\libbase64-0128d3f5b19dfdd5.rmeta --extern hex_literal=D:\GitHub\PsiStega3\target\debug\deps\hex_literal-c2a5ebba97118bc5.dll --extern image=D:\GitHub\PsiStega3\target\debug\deps\libimage-4ced04878b27868b.rmeta --extern log=D:\GitHub\PsiStega3\target\debug\deps\liblog-8fb4ff5a8792f1ff.rmeta --extern path_absolutize=D:\GitHub\PsiStega3\target\debug\deps\libpath_absolutize-8be0795dd4cccbaa.rmeta --extern rand=D:\GitHub\PsiStega3\target\debug\deps\librand-d4d8e8aa41a6f409.rmeta --extern rand_chacha=D:\GitHub\PsiStega3\target\debug\deps\librand_chacha-db2bde74735ff5f5.rmeta --extern rand_core=D:\GitHub\PsiStega3\target\debug\deps\librand_core-9fff5bc747097dd9.rmeta --extern sha3=D:\GitHub\PsiStega3\target\debug\deps\libsha3-467e9370726ce64d.rmeta --extern simple_logger=D:\GitHub\PsiStega3\target\debug\deps\libsimple_logger-c35d7138cb0dd367.rmeta`
[glib-sys 0.14.0] cargo:rerun-if-env-changed=GLIB_2.0_NO_PKG_CONFIG
[glib-sys 0.14.0] cargo:rerun-if-env-changed=PKG_CONFIG_x86_64-pc-windows-msvc
[glib-sys 0.14.0] cargo:rerun-if-env-changed=PKG_CONFIG_x86_64_pc_windows_msvc
[glib-sys 0.14.0] cargo:rerun-if-env-changed=HOST_PKG_CONFIG
[glib-sys 0.14.0] cargo:rerun-if-env-changed=PKG_CONFIG
[glib-sys 0.14.0] cargo:rerun-if-env-changed=PKG_CONFIG_PATH_x86_64-pc-windows-msvc
[glib-sys 0.14.0] cargo:rerun-if-env-changed=PKG_CONFIG_PATH_x86_64_pc_windows_msvc
[glib-sys 0.14.0] cargo:rerun-if-env-changed=HOST_PKG_CONFIG_PATH
[glib-sys 0.14.0] cargo:rerun-if-env-changed=PKG_CONFIG_PATH
[glib-sys 0.14.0] cargo:rerun-if-env-changed=PKG_CONFIG_LIBDIR_x86_64-pc-windows-msvc
[glib-sys 0.14.0] cargo:rerun-if-env-changed=PKG_CONFIG_LIBDIR_x86_64_pc_windows_msvc
[glib-sys 0.14.0] cargo:rerun-if-env-changed=HOST_PKG_CONFIG_LIBDIR
[glib-sys 0.14.0] cargo:rerun-if-env-changed=PKG_CONFIG_LIBDIR
[glib-sys 0.14.0] cargo:rerun-if-env-changed=PKG_CONFIG_SYSROOT_DIR_x86_64-pc-windows-msvc
[glib-sys 0.14.0] cargo:rerun-if-env-changed=PKG_CONFIG_SYSROOT_DIR_x86_64_pc_windows_msvc
[glib-sys 0.14.0] cargo:rerun-if-env-changed=HOST_PKG_CONFIG_SYSROOT_DIR
[glib-sys 0.14.0] cargo:rerun-if-env-changed=PKG_CONFIG_SYSROOT_DIR
[glib-sys 0.14.0] cargo:warning=Could not run `"pkg-config" "--libs" "--cflags" "glib-2.0" "glib-2.0 >= 2.66"`
[glib-sys 0.14.0] The pkg-config command could not be found.
[glib-sys 0.14.0]
[glib-sys 0.14.0] Most likely, you need to install a pkg-config package for your OS.
[glib-sys 0.14.0]
[glib-sys 0.14.0] If you've already installed it, ensure the pkg-config command is one of the
[glib-sys 0.14.0] directories in the PATH environment variable.
[glib-sys 0.14.0]
[glib-sys 0.14.0] If you did not expect this build to link to a pre-installed system library,
[glib-sys 0.14.0] then check documentation of the glib-sys crate for an option to
[glib-sys 0.14.0] build the library from source, or disable features or dependencies
[glib-sys 0.14.0] that require pkg-config.
The following warnings were emitted during compilation:

warning: Could not run `"pkg-config" "--libs" "--cflags" "glib-2.0" "glib-2.0 >= 2.66"`

error: failed to run custom build command for `glib-sys v0.14.0`

Caused by:
  process didn't exit successfully: `D:\GitHub\PsiStega3\target\debug\build\glib-sys-354f0c217b49928d\build-script-build` (exit code: 1)
  --- stdout
  cargo:rerun-if-env-changed=GLIB_2.0_NO_PKG_CONFIG
  cargo:rerun-if-env-changed=PKG_CONFIG_x86_64-pc-windows-msvc
  cargo:rerun-if-env-changed=PKG_CONFIG_x86_64_pc_windows_msvc
  cargo:rerun-if-env-changed=HOST_PKG_CONFIG
  cargo:rerun-if-env-changed=PKG_CONFIG
  cargo:rerun-if-env-changed=PKG_CONFIG_PATH_x86_64-pc-windows-msvc
  cargo:rerun-if-env-changed=PKG_CONFIG_PATH_x86_64_pc_windows_msvc
  cargo:rerun-if-env-changed=HOST_PKG_CONFIG_PATH
  cargo:rerun-if-env-changed=PKG_CONFIG_PATH
  cargo:rerun-if-env-changed=PKG_CONFIG_LIBDIR_x86_64-pc-windows-msvc
  cargo:rerun-if-env-changed=PKG_CONFIG_LIBDIR_x86_64_pc_windows_msvc
  cargo:rerun-if-env-changed=HOST_PKG_CONFIG_LIBDIR
  cargo:rerun-if-env-changed=PKG_CONFIG_LIBDIR
  cargo:rerun-if-env-changed=PKG_CONFIG_SYSROOT_DIR_x86_64-pc-windows-msvc
  cargo:rerun-if-env-changed=PKG_CONFIG_SYSROOT_DIR_x86_64_pc_windows_msvc
  cargo:rerun-if-env-changed=HOST_PKG_CONFIG_SYSROOT_DIR
  cargo:rerun-if-env-changed=PKG_CONFIG_SYSROOT_DIR
  cargo:warning=Could not run `"pkg-config" "--libs" "--cflags" "glib-2.0" "glib-2.0 >= 2.66"`
  The pkg-config command could not be found.

  Most likely, you need to install a pkg-config package for your OS.

  If you've already installed it, ensure the pkg-config command is one of the
  directories in the PATH environment variable.

  If you did not expect this build to link to a pre-installed system library,
  then check documentation of the glib-sys crate for an option to
  build the library from source, or disable features or dependencies
  that require pkg-config.
warning: build failed, waiting for other jobs to finish...
error: build failed

**Comment by sdroege** [comment]
You'll have to check how to teach VS code to set the `PATH` correctly then. There's not much else that can be done :)

**Comment by Hofer-Julian** [comment]
One annoying thing I had was that vscode restores your terminal session including your environment variables.

You can try to disable that by adding this to your settings:

```json
"terminal.integrated.persistentSessionReviveProcess": "never",
````

**Comment by sciguyryan** [comment]
> One annoying thing I had was that vscode restores your terminal session including your environment variables.
> 
> You can try to disable that by adding this to your settings:
> 
> ```json
> "terminal.integrated.persistentSessionReviveProcess": "never",
> ```

Thanks. I actually filed a bug with VS Code since there is no way that this can be intended behaviour.

https://github.com/microsoft/vscode/issues/137736

Luckily, it looks like it has already been fixed for the next release. I think that this issue can be closed since it is not a GTK problem.

Thanks for helping to unpick this one, and thanks for the work around until a fixed build is released!
